<launch> 
 <arg name="global_frame" default ="odom"/>
 <arg name="path_frame" default ="base_link_p"/>
 <arg name="odom_topic" default ="/gazebo/forklift_controllers/odom"/>
 <arg name="cmd_vel_topic" default ="/cmd_vel_raw"/>

 <node pkg="rack_adaptive_dropoff" type="adaptive_dropoff_node" name="adaptive_dropoff_node" output = "screen" >
   <param name="global_frame" value="$(arg global_frame)"/>
   <param name="path_frame" value="$(arg path_frame)"/>

<!-- Docking hardcode fake goal -->
    <!-- <param name="approach_ref_x" value="2.5"/>  
    <param name="approach_ref_y" value="-2.3"/>
    <param name="approach_ref_angle" value="-1.57"/>

    <param name="docking_ref_x" value="2.5"/>  
    <param name="docking_ref_y" value="-2.3"/>
    <param name="docking_ref_angle" value="-1.57"/> -->

    <param name="approach_ref_x" value="1.2"/>  
    <param name="approach_ref_y" value="0.3"/>
    <param name="approach_ref_angle" value="0"/>

    <param name="docking_ref_x" value="1.8"/>  
    <param name="docking_ref_y" value="0.3"/>
    <param name="docking_ref_angle" value="0"/>

<!-- Path Planner -->
  <!-- Terminal boundary -->
    <param name="max_starting_vel" value="0.1"/>   
    <param name="max_ending_vel" value="0.3"/>
    <param name="k_ending_lat" value="1.0"/>
    <param name="k_ending_angle" value="0.5"/>
   
  <!-- Max boundary -->
    <param name="max_acc" value="0.5"/>
    <param name="min_highest_acc" value="0.001"/> <!-- Prevent zero max acc -->
    <param name="max_jerk" value="0.2"/>

  <!-- Time condition -->
    <param name="min_t" value="0.01"/>
    <param name="max_t" value="100.0"/>
    <param name="dt" value="0.1"/>   

<!-- Controller -->
    <param name="publish_cmd" value="true"/>
    <param name="docking_freq" value="10.0"/>

    <param name="alpha" value="1.0"/> <!-- low pass filter param -->

    <param name="tau" value="1.5"/>

    <param name="use_velocity_penalty" value="false"/>
    <param name="predict_time" value="3.0"/> <!-- Predict time for cost function checking -->
    <param name="velocity_penalty_dis_thres" value="0.25"/> <!-- Distance threshold to penatrate the velocity (m) -->
    <param name="obstacle_thresh" value="0.05"/> <!-- Distance threshold to stop the AGV (m) -->

  <!-- Control limitation parameters -->
    <param name="max_angular_vel" value="0.5"/> 
    <param name="max_pocket_dock_vel" value="0.5"/> 
    <param name="max_pocket_dock_steering" value="1.55"/> 

  <!-- Pure Pursuit parameters -->
    <!-- Steering control limitation -->
    <param name="max_steering" value="1.57"/>     
    <param name="min_steering" value="-1.57"/> 
    <param name="max_steering_speed" value="0.5"/> 
    <param name="min_steering_speed" value="-0.5"/> 

    <!-- Calculate lookahead using heading threshold -->
    <param name="cal_lookahead_using_heading_thres" value="true"/>  
    <param name="heading_thres" value="0.1"/> 

    <!-- Lookahed distance calculation -->
    <param name="look_ahead_time" value="3.6"/>     
    <param name="min_look_ahead_dis" value="0.2"/>  
    <param name="max_look_ahead_dis" value="0.5"/> 
    <param name="re_cal_lookahead_dis" value="true"/>
    <param name="use_point_interpolate" value="false"/> 

    <!-- Reference angle regulation -->
    <param name="adaptive_ref_angle" value="false"/> 
    <param name="use_ref_angle_from_path" value="false"/> 
    <param name="goal_correct_yaw" value="0.1"/> 

    <!-- Lateral control supportive params -->
    <param name="kp" value="2.0"/>     
    <param name="ki" value="0.1"/> 
    <param name="k_lat" value="1.0"/> 
    <param name="k_angle" value="1.0"/> 

  <!-- Fuzzy controller parameters -->
    <param name="fuzzy_lookahead_dis" value="1.0"/>      <!-- ZERO value for output velocity --> 
    <param name="limit_sp_curve" value="true"/> 
    <!-- Control language variables -->
    <param name="min_vel" value="0.01"/>      
    <param name="output_VS" value="0.05"/>    
    <param name="output_S" value="0.1"/>     
    <param name="output_M" value="0.2"/>     
    <param name="max_vel" value="0.3"/>       

    <!-- Distance input language variables -->
    <param name="min_dis" value="0.03"/>      <!-- SMALL distance for input -->
    <param name="goal_M_b" value="0.3"/>      <!-- MEDIUM distance for input -->
    <param name="max_dis" value="0.6"/>      <!-- BIG distance for input -->

    <!-- Path curve input language variables -->
    <param name="min_curve" value="-1.0"/>  <!-- NEGATIVE BIG steering for input -->
    <param name="curve_NS_b" value="-0.2"/>   <!-- NEGATIVE SMALL steering for input -->
    <param name="curve_Z_b" value="0.0"/>     <!-- ZERO steering for input -->
    <param name="curve_PS_b" value="0.2"/>    <!-- POSITIVE SMALL steeringfor input -->
    <param name="max_curve" value="1.0"/>   <!-- POSITIVE BIG steering for input -->

    <!-- Offset backward motion -->
    <param name="backward_offset" value="-0.0"/> <!-- offset velocity used to switch to reverse movement.
                                                  should be small negative -->
    <param name="max_linear_acc" value="0.5"/>
    <param name="min_linear_acc" value="-0.5"/>

  <remap from="/cmd_vel" to = "$(arg cmd_vel_topic)"/>
  <remap from="/gazebo/forklift_controllers/odom" to = "$(arg odom_topic)"/>
 </node>
</launch>
